lib/broccoli/ember-app.js

```diff
@@ -682,34 +682,46 @@ class EmberApp {
     return this._addonTreesFor(type).map(addonBundle => addonBundle.tree);
   }
 
-  _getDefaultCompiler(type) {
-    // since we are now using the app's registry to compile addon code,
-    // we could get into the case where the app has multiple compilers (coffeescript, emblem, etc.)
-    // operating on addon code where there weren't previously.
-    // This prevents that by using the default app compiler only.
+  _getDefaultPluginForType(type) {
     let plugins = this.registry.load(type);
-    return plugins[plugins.length - 1];
+    let defaultsForType = plugins.filter(plugin => plugin.isDefaultForType);
+
+    if (defaultsForType.length > 1) {
+      throw new Error(
+        `There are multiple preprocessor plugins marked as default for '${type}': ${defaultsForType.map(p => p.name).join(', ')}`
```

Comment:
`p.name` actually _does not_ reflect reality, we don't require pushing `name` property when we register preprocessors:

```javascript
registry.add('template', {
  isDefaultForType: true,
  ext: 'hbs',
  toTree(tree) { return tree; }
});
```

Reply:
we could change the way we put together `precompiledSource` from `addonBundles` and get the `name` that way, or could just ask for additional `name` property during registration (the first option is more work)

Reply:
Seems like most things use a name already (ember-cli-babel does for sure)...

Reply:
We will need to patch `ember-cli-babel` and `ember-cli-htmlbars` either way (as well as other preprocessors)